File: package/ios/ShapeDescriptor+RenderVersion.swift
/// Render versions for shape overlay descriptors.
///
/// Overlay controllers keep the version of every shape they have shown so a
/// descriptor that is sent again unchanged costs one hash instead of a native
/// remove-and-add. Geometry and style are versioned separately: a geometry
/// change needs a new SDK overlay, a style-only change is applied in place.
extension Array where Element == Coordinate {
  func hashGeometry(into hasher: inout Hasher) {
    hasher.combine(count)
    for coordinate in self {
      hasher.combine(coordinate.latitude)
      hasher.combine(coordinate.longitude)
    }
  }
}

extension PolylineDescriptor {
  func geometryVersion() -> Int {
    var hasher = Hasher()
    coordinates.hashGeometry(into: &hasher)
    return hasher.finalize()
  }

  func styleVersion() -> Int {
    var hasher = Hasher()
    hasher.combine(strokeColor)
    hasher.combine(strokeWidth)
    hasher.combine(tappable)
    return hasher.finalize()
  }
}

extension PolygonDescriptor {
  func geometryVersion() -> Int {
    var hasher = Hasher()
    coordinates.hashGeometry(into: &hasher)
    return hasher.finalize()
  }

  func styleVersion() -> Int {
    var hasher = Hasher()
    hasher.combine(fillColor)
    hasher.combine(strokeColor)
    hasher.combine(strokeWidth)
    hasher.combine(tappable)
    return hasher.finalize()
  }
}

extension CircleDescriptor {
  func geometryVersion() -> Int {
    var hasher = Hasher()
    hasher.combine(center.latitude)
    hasher.combine(center.longitude)
    hasher.combine(radius)
    return hasher.finalize()
  }

  func styleVersion() -> Int {
    var hasher = Hasher()
    hasher.combine(fillColor)
    hasher.combine(strokeColor)
    hasher.combine(strokeWidth)
    hasher.combine(tappable)
    return hasher.finalize()
  }
}

/// Geometry and style versions of a shown shape overlay, kept per overlay id.
struct ShapeRenderVersion: Equatable {
  let geometry: Int
  let style: Int
}

extension PolylineDescriptor {
  func renderVersion() -> ShapeRenderVersion {
    ShapeRenderVersion(geometry: geometryVersion(), style: styleVersion())
  }
}

extension PolygonDescriptor {
  func renderVersion() -> ShapeRenderVersion {
    ShapeRenderVersion(geometry: geometryVersion(), style: styleVersion())
  }
}

extension CircleDescriptor {
  func renderVersion() -> ShapeRenderVersion {
    ShapeRenderVersion(geometry: geometryVersion(), style: styleVersion())
  }
}

File: package/android/src/main/java/com/margelo/nitro/nitromaps/ShapeDescriptor+RenderVersion.kt
package com.margelo.nitro.nitromaps

/**
 * Render versions for shape overlay descriptors.
 *
 * Overlay controllers keep the version of every shape they have shown so a
 * descriptor that is sent again unchanged costs one hash instead of a native
 * remove-and-add, and a changed one is updated in place.
 */
private fun Array<Coordinate>.geometrySignature(): Long {
  var hash = size.toLong()
  for (coordinate in this) {
    hash = 1099511628211L * hash + java.lang.Double.doubleToLongBits(coordinate.latitude)
    hash = 1099511628211L * hash + java.lang.Double.doubleToLongBits(coordinate.longitude)
  }
  return hash
}

/** Distinct from a content hash of an empty holes list. */
private const val ABSENT_HOLES_HASH = 0x2_0000_0000L

private fun Array<Array<Coordinate>>?.holesSignature(): Long {
  if (this == null) {
    return ABSENT_HOLES_HASH
  }

  var hash = size.toLong()
  for (hole in this) {
    hash = 1099511628211L * hash + hole.geometrySignature()
  }
  return hash
}

internal fun PolylineDescriptor.renderVersion(): Long =
  renderSignature(
    "polyline",
    id,
    coordinates.geometrySignature(),
    strokeColor,
    strokeWidth,
    zIndex,
    tappable,
  )

internal fun PolygonDescriptor.renderVersion(): Long =
  renderSignature(
    "polygon",
    id,
    coordinates.geometrySignature(),
    holes.holesSignature(),
    fillColor,
    strokeColor,
    strokeWidth,
    zIndex,
    tappable,
  )

internal fun CircleDescriptor.renderVersion(): Long =
  renderSignature(
    "circle",
    id,
    center.latitude,
    center.longitude,
    radius,
    fillColor,
    strokeColor,
    strokeWidth,
    tappable,
  )
